tests/testmplex.nim

```diff
@@ -594,6 +594,347 @@ suite "Mplex":
+        dialStreams.add((await mplexDial.newStream()))
+
+      mplexHandle.cancel()
+      await allFuturesThrowing(
+        allFinished(
+          (dialStreams & listenStreams)
+          .mapIt( it.join() )))
+
+      checkTracker(LPChannelTrackerName)
+
+      await conn.close()
+      await mplexDialFut
+      await allFuturesThrowing(
+        transport1.close(),
+        transport2.close())
+      await listenFut
+
+    waitFor(testNewStream())
+
+  test "e2e - closing dialing connection should close both ends":
+    proc testNewStream() {.async.} =
+      let ma: MultiAddress = Multiaddress.init("/ip4/0.0.0.0/tcp/0").tryGet()
+
+      var listenStreams: seq[Connection]
+      proc connHandler(conn: Connection) {.async, gcsafe.} =
+        let mplexListen = Mplex.init(conn)
+        mplexListen.streamHandler = proc(stream: Connection)
+          {.async, gcsafe.} =
+            listenStreams.add(stream)
+            await stream.join()
+
+        await mplexListen.handle()
+        await mplexListen.close()
+
+      let transport1 = TcpTransport.init()
+      let listenFut = await transport1.listen(ma, connHandler)
+
+      let transport2: TcpTransport = TcpTransport.init()
+      let conn = await transport2.dial(transport1.ma)
+
+      let mplexDial = Mplex.init(conn)
+      let mplexDialFut = mplexDial.handle()
+      var dialStreams: seq[Connection]
+      for i in 0..9:
+        dialStreams.add((await mplexDial.newStream()))
+
+      await conn.close()
+      await allFuturesThrowing(
+        allFinished(
+          (dialStreams & listenStreams)
+          .mapIt( it.join() )))
+
+      checkTracker(LPChannelTrackerName)
+
+      await conn.close()
+      await mplexDialFut
+      await allFuturesThrowing(
+        transport1.close(),
+        transport2.close())
+      await listenFut
+
+    waitFor(testNewStream())
+
+  test "e2e - canceling listening connection should close both ends":
+    proc testNewStream() {.async.} =
+      let ma: MultiAddress = Multiaddress.init("/ip4/0.0.0.0/tcp/0").tryGet()
+
+      var listenConn: Connection
+      var listenStreams: seq[Connection]
+      proc connHandler(conn: Connection) {.async, gcsafe.} =
+        listenConn = conn
+        let mplexListen = Mplex.init(conn)
+        mplexListen.streamHandler = proc(stream: Connection)
+          {.async, gcsafe.} =
+            listenStreams.add(stream)
+            await stream.join()
+
+        await mplexListen.handle()
+        await mplexListen.close()
+
+      let transport1 = TcpTransport.init()
+      let listenFut = await transport1.listen(ma, connHandler)
+
+      let transport2: TcpTransport = TcpTransport.init()
+      let conn = await transport2.dial(transport1.ma)
+
+      let mplexDial = Mplex.init(conn)
+      let mplexDialFut = mplexDial.handle()
+      var dialStreams: seq[Connection]
+      for i in 0..9:
+        dialStreams.add((await mplexDial.newStream()))
+
+      await listenConn.close()
+      await allFuturesThrowing(
```

Comment:
hm, I wonder.. this could probably be written as `mapIt(dial&listen, await it.join())`

Reply:
Sure, but I honestly prefer the more consistent method call syntax.

Reply:
sure, it's a bit pointless to do double "all" though.. ie `allFuturesThrowing(mapit(...))` does the same at the end of the day

Reply:
ah, yeah - you're right 👍